File: Problems/0338. (E) Counting Bits/Python3/solution.py
class Solution:
  def countBits(self, n: int) -> List[int]:
    # memoize previous results
    
    ret = [0] * (n+1)
    for i in range(n+1):
      ret[i] = ret[i//2] + i%2 if i else 0
    
    return ret

File: Problems/0338. (E) Counting Bits/Java/solution.java
class Solution {
  public int[] countBits(int n) {
    // memoize previous results

    int[] ret = new int[n+1];
    for (int i = 0; i <= n; i++) {
      // bit manipulation faster than modulo or division
      ret[i] = ret[i>>1] + (i & 1);
    }
    return ret;
  }
}
